### PR TITLE
[ML] Improve bad input handling in distribution models

### PR DESCRIPTION
**Pull request by tveasey**
In particular, it more reliably stops non-finite inputs polluting the model.

### CODE REVIEW

File: lib/maths/CLogNormalMeanPrecConjugate.cc [hunk 1]
```diff
@@ -577,10 +577,11 @@ class CLogSampleSquareDeviation : core::CNonCopyable {
         result = 0.0;
         for (std::size_t i = 0u; i < m_Samples.size(); ++i) {
             double residual = m_Samples[i];
-            if (residual <= 0.0) {
+            double n = maths_t::countForUpdate(m_Weights[i]);
+            if (residual <= x || !CMathsFuncs::isFinite(residual) ||
```
Comment (droberts195) [inline lib/maths/CLogNormalMeanPrecConjugate.cc:581]: Was it intentional to change `<= 0.0` to `<= x`?

Comment (tveasey) [reply lib/maths/CLogNormalMeanPrecConjugate.cc:581]: Yes. This is actually was actually a bug I spotted while working on this. It causes us to discard a small number of sample we shouldn't be.

Comment (tveasey) [reply lib/maths/CLogNormalMeanPrecConjugate.cc:581]: Although, related tests have failed, so let me double check this.

Comment (tveasey) [reply lib/maths/CLogNormalMeanPrecConjugate.cc]: I got the new condition wrong as well :(. Corrected.

**Pull request merged**